File: app/src/main/java/com/example/androidscript/Activities/FGO/FGOButtonAdapter.java
package com.example.androidscript.Activities.FGO;

import androidx.annotation.NonNull;

import com.example.androidscript.UITemplate.BlockAdapter;
import com.example.androidscript.UITemplate.ButtonAdapter;

import java.util.Arrays;
import java.util.Vector;

public class FGOButtonAdapter extends ButtonAdapter {
    private final Vector<Vector<String>> BlockContent;
    private static final int insertPosition = 1;
    BlockAdapter.updater onInsert;

    public FGOButtonAdapter(Vector<Vector<String>> _BlockContent, Vector<String> _ButtonText, BlockAdapter.updater _onInsert) {
        super(_ButtonText);
        this.ButtonText = _ButtonText;
        this.BlockContent = _BlockContent;
        this.onInsert = _onInsert;
    }

    @Override
    public void onBindViewHolder(@NonNull ButtonViewHolder holder, int position) {
        switch (ButtonText.get(position)) {
            case "ServantSkill":
                holder.button.setText("從者技能");
                holder.button.setOnClickListener(v -> {
                    BlockContent.insertElementAt(new Vector<>(Arrays.asList(FGOEditor.SkillBlock)), insertPosition);
                    onInsert.insert();
                });
                break;
            case "SelectCard":
                holder.button.setText("自動選卡");
                holder.button.setOnClickListener(v -> {
                    BlockContent.insertElementAt(new Vector<>(Arrays.asList(FGOEditor.NoblePhantasmsBlock)), insertPosition);
                    onInsert.insert();
                });
                break;
            case "CraftSkill":
                holder.button.setText("御主技能");
                holder.button.setOnClickListener(v -> {
                    BlockContent.insertElementAt(new Vector<>(Arrays.asList(FGOEditor.CraftSkillBlock)), insertPosition);
                    onInsert.insert();
                });
                break;
            default:
                throw new RuntimeException("Unrecognized button!");
        }
    }
}

File: app/src/main/java/com/example/androidscript/activities/fgo/FGOButtonAdapter.kt
package com.example.androidscript.activities.fgo

import com.example.androidscript.activities.template.BlockAdapter.Updater
import com.example.androidscript.activities.template.ButtonAdapter
import java.util.*

class FGOButtonAdapter(
    _blockContent: Vector<Vector<String>>,
    _buttonText: Vector<String>,
    _onInsert: Updater
) : ButtonAdapter(_buttonText) {
    private val blockContent: Vector<Vector<String>>
    private var onInsert: Updater
    override fun onBindViewHolder(holder: ButtonViewHolder, position: Int) {
        when (buttonText[position]) {
            "ServantSkill" -> {
                holder.button.text = "從者技能" // TODO extract string resources.
                holder.button.setOnClickListener {
                    blockContent.insertElementAt(
                        Vector(listOf(*FGOEditor.skillBlockData)),
                        insertPosition
                    )
                    onInsert.insert()
                }
            }
            "SelectCard" -> {
                holder.button.text = "自動選卡"
                holder.button.setOnClickListener {
                    blockContent.insertElementAt(
                        Vector(listOf(*FGOEditor.noblePhantasmBlockData)),
                        insertPosition
                    )
                    onInsert.insert()
                }
            }
            "CraftSkill" -> {
                holder.button.text = "御主技能"
                holder.button.setOnClickListener {
                    blockContent.insertElementAt(
                        Vector(listOf(*FGOEditor.craftSkillBlockData)),
                        insertPosition
                    )
                    onInsert.insert()
                }
            }
            else -> throw RuntimeException("Unrecognized button!")
        }
    }

    companion object {
        private const val insertPosition = 1
    }

    init {
        buttonText = _buttonText
        blockContent = _blockContent
        onInsert = _onInsert
    }
}